### PR TITLE
[Android] Support new file chooser

### PR DESCRIPTION
**Pull request by sunlin-link**
- Use new XWalkFileChooser.
- Request dangerous persmisions dynamically for Marshmallow and above.

BUG=XWALK-7281

### CODE REVIEW

**Comment by lincsoon** [comment]
@fujunwei @xzhan96 Please help to review

**Comment by lincsoon** [comment]
This feature is for Crosswalk-23. If this commit is merged, the plugin will not be compatible with the build prior to Crosswalk-23.

**Comment by xzhan96** [comment]
> This feature is for Crosswalk-23. If this commit is merged, the plugin will not be compatible with the build prior to Crosswalk-23.

Does this info need to be added into commit log?

**Comment by fujunwei** [comment]
Ok, it will be announced at release notes.

**Comment by rakuco** [comment]
Can you explain the source of the incompatibility?

File: platforms/android/src/org/crosswalk/engine/XWalkWebViewEngine.java [hunk 1]
```diff
@@ -161,6 +167,36 @@ public void onResume(boolean multitasking) {
         };
         pluginManager.addService(new PluginEntry("XWalkActivityDelegate", activityDelegatePlugin));
 
+        ArrayList<String> dangerous_permissions = new ArrayList<String>();
```
Comment (fujunwei) [inline platforms/android/src/org/crosswalk/engine/XWalkWebViewEngine.java]: It's better to hold those codes in a function,  and please confirm the feature of choosing file is work finely with App-tools.

File: platforms/android/src/org/crosswalk/engine/XWalkCordovaUiClient.java [hunk 2]
```diff
@@ -172,12 +175,15 @@ public void onPageLoadStopped(XWalkView view, String url, LoadStatus status) {
     // File Chooser
     @Override
     public void openFileChooser(XWalkView view, final ValueCallback<Uri> uploadFile, String acceptType, String capture) {
-        uploadFile.onReceiveValue(null);
+        if (mFileChooser == null) {
+            mFileChooser = new XWalkFileChooser(parentEngine.cordova.getActivity());
```
Comment (fujunwei) [inline platforms/android/src/org/crosswalk/engine/XWalkCordovaUiClient.java]: I suggest get activity internally, and keep compatible

> public Activity getActivity() {
        if (mContext instanceof Activity) {
            return (Activity) mContext;
        }
        return null;
    }

**Comment by lincsoon** [comment]
@rakuco `XWalkFileChooser` is a new class starting with Crosswalk-23.

File: platforms/android/src/org/crosswalk/engine/XWalkCordovaUiClient.java (hunk 2 above)
Comment (lincsoon) [inline platforms/android/src/org/crosswalk/engine/XWalkCordovaUiClient.java]: Now XWalkView can use any context. The developer could pass an application context into it and hold it in an activity. So casting the context inside of XWalkView to activity is not a proper option.

Comment (fujunwei) [inline platforms/android/src/org/crosswalk/engine/XWalkCordovaUiClient.java]: what's the differentiation between Activity and Application Context as as parameter to XWalkView.

Comment (lincsoon) [inline platforms/android/src/org/crosswalk/engine/XWalkCordovaUiClient.java]: The application context cann't be cast to an activity

Comment (fujunwei) [inline platforms/android/src/org/crosswalk/engine/XWalkCordovaUiClient.java]: There is a table of using Context in different scenario https://possiblemobile.com/2013/06/context/,  is it mean that Application Context doesn't suite XWalkView?

Comment (lincsoon) [inline platforms/android/src/org/crosswalk/engine/XWalkCordovaUiClient.java]: On which point do you think  the application context doesn't suite XWalkView?

Comment (lincsoon) [inline platforms/android/src/org/crosswalk/engine/XWalkCordovaUiClient.java]: About the code here, the problem is not the different between the application context and the activity. The point is that the context inside of XWalkView may not be the activity that the developer wants to launch the file chooser.

Comment (fujunwei) [inline platforms/android/src/org/crosswalk/engine/XWalkCordovaUiClient.java]: http://img.blog.csdn.net/20150104183450879
Inflation will be done with the default theme for the system on which you are running, not what’s defined in your application.
Did it verified it worked finely with Application Context?

Comment (fujunwei) [inline platforms/android/src/org/crosswalk/engine/XWalkCordovaUiClient.java]: Could you show this case?
> The point is that the context inside of XWalkView may not be the activity that the developer wants to launch the file chooser.

Comment (lincsoon) [inline platforms/android/src/org/crosswalk/engine/XWalkCordovaUiClient.java]: In which class XWalkView will do the layout inflation?

Comment (lincsoon) [inline platforms/android/src/org/crosswalk/engine/XWalkCordovaUiClient.java]: > Could you show this case?

I don't have the code here. I just say that we can support this case.

Comment (fujunwei) [inline platforms/android/src/org/crosswalk/engine/XWalkCordovaUiClient.java]: How about implementing with Activity context by default and adding a new API for special requirements.

Comment (lincsoon) [inline platforms/android/src/org/crosswalk/engine/XWalkCordovaUiClient.java]: Could you explain more about implementinig with Activity by default? And which class to add the new API?

**Comment by lincsoon** [comment]
@fujunwei Updated. The change for app tools is in another patch of crosswalk repo.

> It's better to hold those codes in a function, and please confirm the feature of choosing file is work finely with App-tools.

**Comment by marcuspridham** [comment]
The Chromium SelectFileDialog requests the 6.0 permissions there.
https://github.com/crosswalk-project/chromium-crosswalk/blob/master/ui/android/java/src/org/chromium/ui/base/SelectFileDialog.java#L120

Should this change go in the XWalkFileChooser instead?  With the current pull request, the Crosswalk plugin is going to spam the user for camera permissions and external storage when they might not ever even use the file chooser.

**Comment by lincsoon** [comment]
You're right! We are going to change the timing of requesting dangerous permissions to trigger it only when the file chooser is to be launched. Though this pull request is not supposed to be merged soon.

**Comment by lincsoon** [comment]
Updated. Requests permissions only when the file chooser is going to be launched.

**Comment by lincsoon** [comment]
@fujunwei Please remember to merge this pull request when the dependent crosswalk's version is 23.

**Comment by dpogue** [comment]
This doesn't seem to open the file picker for the correct acceptType. If my HTML is `<input type=file accept="image/*">` it only offers Camera and Documents, not the Photos app.

**Comment by fujunwei** [comment]
Yes,  capture="true" attribute need to be added if you want only capture image with camera.

**Comment by dpogue** [comment]
Yeah, I've confirmed my issue is unrelated to this PR's code, and I've filed XWALK-7451 to track it separately.

**Comment by ascorbic** [comment]
Is there a timeline for merging this? I'm having issues with the old file chooser (photos from the camera are deleted before they can be read), and would rather not spend ages tracking down the problem if this fixes it.

**Comment by lincsoon** [comment]
This patch has to wait for Crosswalk-23, but the release date of it is uncertain.

**Comment by sunlin-link** [comment]
@fujunwei Updated. Please help to review

**Pull request merged**